README.md
# UNIVERSAL::Object

[![Build Status](https://travis-ci.org/stevan/p5-UNIVERSAL-Object.svg?branch=master)](https://travis-ci.org/stevan/p5-UNIVERSAL-Object)

A useful base class

[![CPAN version](https://badge.fury.io/pl/UNIVERSAL-Object.svg)](https://metacpan.org/pod/UNIVERSAL-Object)

# Copyright and License

This software is copyright (c) 2017 by Stevan Little.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Perl extension UNIVERSAL::Object

0.15
    - improve some error messages
    - Immutable objects now correctly work with overloaded
      objects by accessing the overload::StrVal instead of
      the overloaded string
    - If your REPR type is a HASH then we will lock the set
      of keys to be those of your set of SLOTS. This should
      not break anything if you were not already violating
      encapsulation somehow ;)

0.14 2018-01-25
    - adding in Travis CI integration
        - thanks to JJ Merelo for this
    - slots are now sorted by name before we attempt to
      populate the instance, this should make some errors
      more predictable
        - thanks to Tony Inkster for this
    - Avoid the stringy eval stuff around mro/MRO::Compat
        - thanks to Graham Knopp for this

0.13 2017-10-30
    - switched back to Carp::confess from Carp::cluck, the
      stack trace is vital until this module matures and
      the error messages improve

0.12 2017-10-18
    - UNIVERSAL::Object::Immutable now allows for
      alteration of the instance during `BUILD` by
      delaying locking of the instance until the
      very last moment before `new` returns
        - updated docs for this
    - UNIVERSAL::Object::Immutable now supports
      CODE ref based instances
        - updated docs and tests for this
    - added test to demonstrate that `%HAS` is not a
      requirement and (ab)using SLOTS method works
      too

0.11 2017-06-19
    - fix a test that is causing CPANTS to fail, sorry

0.10 2017-05-12
    - Improve the content of the errors produced by
      this module so that it is clearer where the
      errors are coming from without having to view
      the stack trace. If that is needed there is
      the Carp::Always module.
        - tweaked the tests to match the new errors

0.09 2017-05-08
    - add support to UNIVERSAL::Object::Immutable
      to handle additional REPR types
        - handles SCALAR, REF, ARRAY and HASH
        - added tests for this

0.08 2017-04-07
    - fixing spelling mistakes and bad POD formatting

0.07 2017-04-06
    - tightened up the perl version requirements in the
      modules as well as the config
    - adding UNIVERSAL::Object::Immutable for creating
      immutable instances
        - added tests and docs for this
        - NOTE: this class requires v5.8 or greater

0.06 2017-02-08
    - cleaning up some test failures and some sloppy
      test writing on my part
        - *sigh*

0.05 2017-02-07
    - the REPR method now takes the $proto which results
      from BUILDARGS in case it needs to consult it in
      the creation of the representation
    - switch to using Carp::croak instead of die
      it is, and always has been, in core.
    - added some tests to show how to inherit from
      classes which do not use UNIVERSAL::Object
    - added new tests for non-HASH ref based instance

0.04 2016-12-29
    - only require MRO::Compat for perls before 5.10
        - thanks to Dan Book for the dist.ini fix

0.03 2016-12-27
    - added a `REPR` method which returns the structure
      to be used as an instance
      	- added docs and tests
    - added a `BLESS` method which handles blessing and
      calling the `CREATE` method to construct the
      instance
      	- added docs and tests

0.02 2016-12-19
    - fixing an issue in a test, looks like if you call
      a package namespace `SUPER` it causes issues.

0.01 2016-12-18
    - Initial release to an unsuspecting world


lib/UNIVERSAL/Object/Immutable.pm
package UNIVERSAL::Object::Immutable;
# ABSTRACT: Another useful base class
use 5.008;
use strict;
use warnings;

use Carp         ();
use overload     ();
use Hash::Util   ();
use Scalar::Util ();

use UNIVERSAL::Object;

our $VERSION   = '0.15';
our $AUTHORITY = 'cpan:STEVAN';

our @ISA; BEGIN { @ISA = ('UNIVERSAL::Object') }

sub new {
    my $class = shift;
    my $self  = $class->SUPER::new( @_ );
    my $repr  = overload::StrVal($self);

    if ( $repr =~ /\=HASH\(0x/ ) {
        Hash::Util::lock_hash( %$self );
    }
    elsif ( $repr =~ /\=ARRAY\(0x/ ) {
        Internals::SvREADONLY( @$self, 1 );
    }
    elsif ( $repr =~ /\=SCALAR\(0x/ or $repr =~ /\=REF\(0x/ ) {
        Internals::SvREADONLY( $$self, 1 );
    }
    elsif ( $repr =~ /\=CODE\(0x/ ) {
        # NOTE: do nothing here, because – ignoring
        # closures – CODE refs are immutable anyway
    }
    else {
        Carp::confess('Invalid BLESS args for '.Scalar::Util::blessed($self).', unsupported REPR type ('.Scalar::Util::reftype($self).')');
    }

    return $self;
}

1;

__END__

=pod

=head1 SYNOPSIS

    # used exactly as UNIVERSAL::Object is used
    our @ISA = ('UNIVERSAL::Object::Immutable');

=head1 DESCRIPTION

You can use this class in the same manner that you would use
L<UNIVERSAL::Object>, the only difference is that the instances
created will be immutable.

=head2 When are instances made Immutable?

Obviously we need to create and initialize the instance before
we make it immutable, it is only after that when we want it to be
immutable. This means that if you need to munge values, or build
values based on the contents of other slots, you should do this
inside a C<BUILD> method, during which the instance will still be
mutable.

=head2 Supported REPR types

This module will attempt to do the right type of locking for
the three main instance types; C<SCALAR>, C<REF>, C<ARRAY>,
C<HASH> and C<CODE>, all other instance types are unsupported
and will throw an error.

=head2 Why Immutability?

Immutable data structures are unable to be changed after they are
created. By placing and enforcing the guarantee of immutability,
the users of our class no longer need to worry about a whole class
of problems that arise from mutable state.

=head2 Immutability is semi-viral

Inheriting from an immutable class will make your subclass also
immutable. This is by design.

When an immutable instance references other data structures, they
are not made immutable automatically. This too is by design.

This means that given the following class:

    package Person {
        use strict;
        use warnings;

        our @ISA = ('UNIVERSAL::Object::Immutable');
        our %HAS = (
            given_names => sub { +[] },
            family_name => sub {  '' },
        );
    }

    package Employee {
        use strict;
        use warnings;

        our @ISA = ('Person');
        our %HAS = (
            %Person::HAS,
            job_title => sub { '' },
        );
    }

Any of the following lines would cause an error about modification
of a read-only value because we are trying to change the values
inside the hashes or add new values.

    my $e = Employee->new;
    $e->{family_name} = 'little';
    $e->{family_name} .= 'little';
    $e->{given_names} = [ 'stevan', 'calvert' ];
    $e->{job_title} = 'developer';
    $e->{misspelled_key} = 0;

However, the following would not cause an error because the C<ARRAY>
reference stored in the C<given_names> slot is not itself read-only
and so can be modified in this way.

    my $e = Employee->new;
    push @{ $e->{given_names} } => 'stevan', 'calvert';

=head2 Immutability after the fact

When this class is used at the root of an object hierarchy, all the
subclasses will be immutable. However, if you wish to make an immutable
subclass of a non-immutable class, then you have two choices.

B<NOTE:> both of the examples below assume that there is not a locally
defined C<new> in your class.

=over 4

=item Multiple Inheritance

Using multiple inheritance, and putting this class first in the list,
we can be sure that the expected version of C<new> is used.

    our @ISA = (
        'UNIVERSAL::Object::Immutable',
        'My::Super::Class'
    );

=item Role Composition

Using the role composition facilities in the L<MOP> package will result
in C<new> being aliased into the consuming package and therefore have
the same effect as the multiple inheritance.

    our @ISA  = ('My::Super::Class');
    our @DOES = ('UNIVERSAL::Object::Immutable');
    # MOP::Util can be used to performs the role composition

=back

=head2 Compatibility Note

This class requires Perl 5.8.0 or later, this is because it
depends on the L<Hash::Util> module and the C<Internals::SvREADONLY>
feature, both of which were first introduced in that version of Perl.
Since this is an optional component, we have not bumped the version
requirement for the entire distribution, only for this module
specifically.

=cut






lib/UNIVERSAL/Object.pm
package UNIVERSAL::Object;
# ABSTRACT: A useful base class
use 5.008;
use strict;
use warnings;

use Carp       ();
use Hash::Util ();

our $VERSION   = '0.15';
our $AUTHORITY = 'cpan:STEVAN';

BEGIN { $] >= 5.010 ? require mro : require MRO::Compat }

sub new {
    my $class = shift;
       $class = ref $class if ref $class;

    my $proto = $class->BUILDARGS( @_ );

    Carp::confess('BUILDARGS must return a HASH reference, not '.$proto)
        unless $proto && ref $proto eq 'HASH';

    my $self = $class->BLESS( $proto );

    Carp::confess('BLESS must return a blessed reference, not '.$self)
        unless defined $self && UNIVERSAL::isa( $self, 'UNIVERSAL' );

    $self->can('BUILD') && UNIVERSAL::Object::Util::BUILDALL( $self, $proto );

    return $self;
}

sub BUILDARGS {
    my $class = shift;
    if ( scalar @_ == 1 && ref $_[0] ) {
        Carp::confess('Invalid BUILDARGS args for '.$class.', expected a HASH reference but got a '.$_[0])
            unless ref $_[0] eq 'HASH';
        return +{ %{ $_[0] } };
    }
    else {
        Carp::confess('Invalid BUILDARGS args for '.$class.', expected an even sized list, but got '.(scalar @_).' element(s) instead')
            unless ((scalar @_) % 2) == 0;
        return +{ @_ };
    }
}

sub BLESS {
    my $class = $_[0];
       $class = ref $class if ref $class;
    my $proto = $_[1];

    Carp::confess('Invalid BLESS args for '.$class.', You must specify an instance prototype as a HASH ref')
        unless defined $proto && ref $proto eq 'HASH';

    my $instance = $class->CREATE( $proto );

    Carp::confess('CREATE must return a reference to bless, not '.$instance)
        unless defined $instance && ref $instance;

    my $repr = ref   $instance;
    my $self = bless $instance => $class;

    # So,... for HASH based instances we'll
    # lock the set of keys so as to prevent
    # typos and other such silliness, if
    # you use other $repr types, you are
    # on your own, ... sorry ¯\_(ツ)_/¯
    if ( $repr eq 'HASH' ) {
        my %slots = $self->SLOTS;
        Hash::Util::lock_keys( %$self, keys %slots );
    }

    return $self;
}

sub CREATE {
    my $class = $_[0];
       $class = ref $class if ref $class;
    my $proto = $_[1];

    my $self  = $class->REPR( $proto );
    my %slots = $class->SLOTS;

    # NOTE:
    # We could check the return values of SLOTS
    # and REPR, but they might change and so it
    # is not something we would always know.
    # - SL

    $self->{ $_ } = exists $proto->{ $_ }
        ? $proto->{ $_ }
        : $slots{ $_ }->( $self, $proto )
            foreach sort keys %slots;

    return $self;
}

sub REPR () { +{} }

sub SLOTS {
    my $class = $_[0];
       $class = ref $class if ref $class;
    no strict   'refs';
    no warnings 'once';
    return %{$class . '::HAS'};
}

sub DESTROY {
    my $self = $_[0];
    $self->can('DEMOLISH') && UNIVERSAL::Object::Util::DEMOLISHALL( $self );
    return;
}

## Utils

sub UNIVERSAL::Object::Util::BUILDALL {
    my $self  = $_[0];
    my $proto = $_[1];
    foreach my $super ( reverse @{ mro::get_linear_isa( ref $self ) } ) {
        my $fully_qualified_name = $super . '::BUILD';
        $self->$fully_qualified_name( $proto )
            if defined &{ $fully_qualified_name };
    }
}

sub UNIVERSAL::Object::Util::DEMOLISHALL {
    my $self = $_[0];
    foreach my $super ( @{ mro::get_linear_isa( ref $self ) } ) {
        my $fully_qualified_name = $super . '::DEMOLISH';
        $self->$fully_qualified_name()
            if defined &{ $fully_qualified_name };
    }
}

1;

__END__

=pod

=head1 SYNOPSIS

    package Person;
    use strict;
    use warnings;
    use UNIVERSAL::Object;

    our @ISA = ('UNIVERSAL::Object');
    our %HAS = (
        name   => sub { die 'name is required' }, # required in constructor
        age    => sub { 0 },                      # w/ default value
        gender => sub {},                         # no default value
    );

    sub name   { $_[0]->{name}   }
    sub age    { $_[0]->{age}    }
    sub gender { $_[0]->{gender} }

    package Employee;
    use strict;
    use warnings;

    our @ISA = ('Person');
    our %HAS = (
        %Person::HAS, # inheritance :)
        job_title => sub { die 'job_title is required' },
        manager   => sub {},
    );

    sub job_title { $_[0]->{job_title} }
    sub manager   { $_[0]->{manager}   }

    # ...

    my $ceo = Employee->new(
        name      => 'Alice',
        job_title => 'CEO',
    );

    my $manager = Employee->new(
        name      => 'Bob',
        job_title => 'Middle Manager',
        manager   => $ceo,
    );

    my $pawn = Employee->new(
        name      => 'Joe',
        job_title => 'Line Worker',
        manager   => $manager,
    );

=head1 DESCRIPTION

This is a simple base class that provides a protocol for object
construction and destruction. It aims to be as simple as possible
while still being complete.

=head1 SLOT MANAGEMENT

One of the key contributions of this module is to provide a
mechanism for declaring the slots that a given class is expected
to have. These are used in the object construction process to ensure
that all slots are created and initialized.

=head2 C<%HAS>

This is a public (C<our>) package variable that contains an entry
for each slot expected in the instance. The key is the slot's name,
while the value is a CODE reference which, when called, will produce
a default value for the slot.

B<NOTE:>
No inheritance of slot definitions is done between classes, this is
left as an exercise to the author, however it is easily accomplished
with the pattern shown above in the L<SYNOPSIS>.

=head2 C<SLOTS ($class)>

This is an accessor method for the C<%HAS> variable in the C<$class>
package.

B<NOTE:>
If you choose to store the slot definitions elsewhere (not in C<%HAS>)
or store them in a different form (not C<< key => CODE >> hash
entries), it is possible to then override this method to return the
expected values in the expected form.

=head1 INSTANCE REPRESENTATION

=head2 C<REPR>

This returns a new HASH reference to use as the instance.

B<NOTE:>
If you wish to use a different instance type of some kind then you
will need to override this method.

B<NOTE:>
The HASH that gets returned from here will eventually be blessed,
which means you are limited to C<tie> or XS C<MAGIC> based solutions
if you want this HASH reference to behave differently.

=head1 CONSTRUCTION PROTOCOL

Once we know the expected slots it is very easy to create a default
constructor. This is the second key contribution of this module, to
provide a consistent and complete protocol for the construction and
destruction of instances.

B<NOTE:>
The method documentation is meant to be read from top to bottom,
in that they are documented in the same order they are called.

=head2 C<new ($class, @args)>

This is the entry point for object construction, from here the
C<@args> are passed into C<BUILDARGS>. The return value of C<new>
should always be a fully constructed and initialized instance.

=head2 C<BUILDARGS ($class, @args)>

This method takes the original C<@args> to the C<new> constructor
and is expected to turn them into a canonical form, which is a
HASH ref of name/value pairs. This form is considered a prototype
candidate for the instance, and what C<BLESS> and subsequently
C<CREATE> expect to receive.

B<NOTE:>
The values in the prototype candidate should be shallow copies of
what was originally contained in C<@args>, but this is not actually
enforced, just suggested to provide better ownership distinctions.

=head2 C<BLESS ($class, $proto)>

This method receives the C<$proto> candidate from C<BUILDARGS> and
from it, ultimately constructs a blessed instance of the class.

This method really has two responsibilities, first is to call
C<CREATE>, passing it the C<$proto> instance. Then it will take
the return value of C<CREATE> and C<bless> it into the C<$class>.

B<NOTE:>
This method will, if the C<REPR> type is a C<HASH> reference,
proceed to lock the set of allowed keys in the C<HASH>. Things
should "Just Work" as long as you do not attempt to access a
slot that you have not defined.

B<NOTE:>
Aside from the above, this method is mostly here to make it
easier to override the C<CREATE> method, which, along with the
C<REPR> method, can be used to change the behavior and/or type
of the instance structure. By keeping the C<bless> work here we
make the work done in C<CREATE> simpler with less mechanics.

=head2 C<CREATE ($class, $proto)>

This method receives the C<$proto> candidate from C<BLESS> and
return from it an unblessed instance structure that C<BLESS> will
then C<bless> into the C<$class>.

First it must call C<SLOTS> (described above), followed by C<REPR>
(also described above) to get both the slot definitions and a newly
minted HASH ref instance. Using these two things, along with the
C<$proto> candidate, we construct a complete blessed instance.
This is accomplished by looping through the list of slots, using
values in the C<$proto> when available, otherwise using the slot
initializers. The final unblessed HASH ref based instance is then
returned.

B<NOTE:>
If you wish to use a different instance type of some kind then you
will need to override this method.

=head2 C<BUILD ($self, $proto)>

The newly blessed instance supplied by C<BLESS> must still be
initialized. We do this by calling all the available C<BUILD> methods
in the inheritance hierarchy in the correct (reverse mro) order.

C<BUILD> is an optional initialization method which receives the
blessed instance as well as the prototype candidate. There are no
restrictions as to what this method can do other then just common
sense.

B<NOTE:>
It is worth noting that because we call all the C<BUILD> methods
found in the object hierarchy, the return values of these methods
are completely ignored.

=head1 DESTRUCTION PROTOCOL

The last thing this module provides is an ordered destruction
protocol for instances.

=head2 C<DEMOLISH ($self)>

This is an optional destruction method, similar to C<BUILD>, all
available C<DEMOLISH> methods are called in the correct (mro) order
by C<DESTROY>.

=head2 C<DESTROY ($self)>

The sole function of this method is to kick off the call to all the
C<DEMOLISH> methods during destruction.

=head1 SEE ALSO

=over 4

=item L<UNIVERSAL::Object::Immutable>

=back

=cut




